libcanard/canard.h

```diff
@@ -1,528 +1,587 @@
- * buffer can be released and re-used by the TX queue.
- * @param ins       Library instance.
- * @param transfer  Pointer to the temporary transfer object. Invalidated upon return.
- */
-typedef void (*CanardOnTransferReception)(CanardInstance* ins, CanardRxTransfer* transfer);
-
-/**
- * This structure provides the usage statistics of the memory pool allocator.
- * It indicates whether the allocated memory is sufficient for the application.
- */
-typedef struct
-{
-    uint16_t capacity_blocks;       ///< Pool capacity in number of blocks
-    uint16_t current_usage_blocks;  ///< Number of blocks that are currently allocated by the library
-    uint16_t peak_usage_blocks;     ///< Maximum number of blocks used since initialization
-} CanardPoolAllocatorStatistics;
+typedef struct CanardInstance CanardInstance;
+typedef uint64_t              CanardMicrosecond;
+typedef uint16_t              CanardPortID;
+typedef uint8_t               CanardNodeID;
+typedef uint8_t               CanardTransferID;
 
-typedef union CanardPoolAllocatorBlock_u
-{
-    char                              bytes[CANARD_MEM_BLOCK_SIZE];
-    union CanardPoolAllocatorBlock_u* next;
-} CanardPoolAllocatorBlock;
-
-typedef struct
+/// Transfer priority level mnemonics per the recommendations given in the UAVCAN Specification.
+typedef enum
 {
-    CanardPoolAllocatorBlock*     free_list;
-    CanardPoolAllocatorStatistics statistics;
-} CanardPoolAllocator;
-
-typedef struct CanardBufferBlock
+    CanardPriorityExceptional = 0,
+    CanardPriorityImmediate   = 1,
+    CanardPriorityFast        = 2,
+    CanardPriorityHigh        = 3,
+    CanardPriorityNominal     = 4,  ///< Nominal priority level should be the default.
+    CanardPriorityLow         = 5,
+    CanardPrioritySlow        = 6,
+    CanardPriorityOptional    = 7,
+} CanardPriority;
+
+/// Transfer kinds as defined by the UAVCAN Specification.
+typedef enum
 {
-    struct CanardBufferBlock* next;
-    uint8_t                   data[];
-} CanardBufferBlock;
+    CanardTransferKindMessage  = 0,  ///< Multicast, from publisher to all subscribers.
+    CanardTransferKindResponse = 1,  ///< Point-to-point, from server to client.
+    CanardTransferKindRequest  = 2,  ///< Point-to-point, from client to server.
+} CanardTransferKind;
+#define CANARD_NUM_TRANSFER_KINDS 3
 
-struct CanardRxState
+/// CAN data frame with an extended 29-bit ID. RTR/Error frames are not used and therefore not modeled here.
+/// CAN frames with 11-bit ID are not used by UAVCAN/CAN and so they are not supported by the library.
+typedef struct
 {
-    struct CanardRxState* next;
-
-    CanardBufferBlock* buffer_blocks;
+    /// For RX frames: reception timestamp.
+    /// For TX frames: transmission deadline.
+    /// The time system may be arbitrary as long as the clock is monotonic (steady).
+    CanardMicrosecond timestamp_usec;
 
-    uint64_t timestamp_usec;
+    /// 29-bit extended ID. The bits above 29-th are zero/ignored.
```

Comment:
```suggestion
    /// 29-bit extended ID. The bits above 29-th are always zero and ignored.
```

Reply:
It was actually a mistake. The higher bits shall be cleared, it is enforced by the API functions. Fixed.